Add user ring buffer support to libbpf-rs

jfernandez opened the pull request:
This change adds support for user ring buffers to libbpf-rs. The user ring buffer is a type of BPF map that allows user space to write data to a ring buffer that can be read by a BPF program.

The motivation for this change is to improve the performance of the `scx_rustland` scheduler in the sched_ext project. We have verified the change by running scx_rustland with the new user ring buffer support.

The `UserRingBuffer` struct is a wrapper around the `Map` type and provides a new set of functions for interacting with the ring buffer. The API closely resembles the libbpf API for user ring buffers. The struct was intentionally made !Send to avoid introducing locking overhead unless necessary. It is expected that the user will handle synchronization if needed.

The `UserRingBufferSample` struct contains a mutable reference to the sample in the ring buffer memory. It is the interface for writing data to the ring buffer. To write to the sample, dereference with `as_mut()` to get a mutable reference to the raw byte slice.

The `user_ring_buffer__discard` function was not exposed because the implementation handles this internally to simplify the API.

The `user_ring_buffer__reserve_blocking` function is not yet implemented and will be added in the future.

Link: https://lore.kernel.org/bpf/20220902234317.2518808-1-void@manifault.com/
Link: https://github.com/sched-ext/scx/tree/main/scheds/rust/scx_rustland
Signed-off-by: Jose Fernandez <josef@netflix.com>
Reviewed-by: Andrea Righi <andrea.righi@canonical.com>

danielocfb commented:
Hi @jfernandez , thanks for the pull request. I think we are going to need a way to test this functionality and then have unit tests. Is this something you can look into?

jfernandez commented:
@danielocfb yes, I can look into that. I wasn't able to easily add basic unit tests because everything calls out to libbpf. If we have integration tests already setup, I can test this behavior there.

jfernandez commented:
@danielocfb I've added a unit test with an accompanying BPF program to validate the new functionality.

danielocfb commented on libbpf-rs/tests/bin/src/user_ringbuf.bpf.c:
```diff
@@ -0,0 +1,47 @@
+	return 0;
+}
+
+SEC("tp_btf/sched_switch")
```
Let's use simple trigger on getpid syscall, as we do elsewhere, to not have everybody do their own thing for no reason?

```patch
diff --git tests/bin/src/user_ringbuf.bpf.c tests/bin/src/user_ringbuf.bpf.c
index debbf2..30d8fd 100644
--- tests/bin/src/user_ringbuf.bpf.c
+++ tests/bin/src/user_ringbuf.bpf.c
@@ -35,9 +35,8 @@ static long user_ringbuf_callback(struct bpf_dynptr *dynptr, void *context)
   return 0;
 }

-SEC("tp_btf/sched_switch")
-int BPF_PROG(handle__sched_switch, bool preempt, struct task_struct *prev,
-       struct task_struct *next)
+SEC("tp/syscalls/sys_enter_getpid")
+int handle__sys_enter_getpid(void *ctx)
 {
   bpf_user_ringbuf_drain(&user_ringbuf, user_ringbuf_callback, NULL, 0);

diff --git tests/bin/user_ringbuf.bpf.o tests/bin/user_ringbuf.bpf.o
index de04d5..40ac6a 100644
Binary files tests/bin/user_ringbuf.bpf.o and tests/bin/user_ringbuf.bpf.o differ
diff --git tests/test.rs tests/test.rs
index 50592a..7b487e 100644
--- tests/test.rs
+++ tests/test.rs
@@ -1205,7 +1205,7 @@ fn test_sudo_object_user_ringbuf() {

     let mut obj = get_test_object("user_ringbuf.bpf.o");
     let prog = obj
-        .prog_mut("handle__sched_switch")
+        .prog_mut("handle__sys_enter_getpid")
         .expect("failed to find program");
     let _link = prog.attach().expect("failed to attach prog");
     let urb_map = obj
@@ -1222,8 +1222,8 @@ fn test_sudo_object_user_ringbuf() {
         .submit(urb_sample)
         .expect("failed to submit sample");

-    // Sleep to ensure the sched_switch event is triggered
-    sleep(Duration::from_millis(10));
+    // Trigger BPF program.
+    let _pid = unsafe { libc::getpid() };

     // At this point, the BPF program should have run and consumed the sample in
     // the user ring buffer, and stored the key/value in the samples map.
```

or something like that?

jfernandez replied:
Changed hook to `sys_enter_getpid`.

danielocfb commented on libbpf-rs/tests/test.rs:
```diff
@@ -1191,6 +1193,55 @@ fn test_sudo_object_ringbuf_with_closed_map() {
+    array.copy_from_slice(&res[..]);
+    assert_eq!(u32::from_ne_bytes(array), value);
+}
+
```
Can we also test error cases such as E2BIG?
```rust
#[test]
fn test_sudo_object_user_ringbuf_reservation_too_big() {
    bump_rlimit_mlock();

    let mut obj = get_test_object("user_ringbuf.bpf.o");
    let prog = obj
        .prog_mut("handle__sys_enter_getpid")
        .expect("failed to find program");
    let _link = prog.attach().expect("failed to attach prog");
    let urb_map = obj
        .map("user_ringbuf")
        .expect("failed to find user ringbuf map");
    let user_ringbuf = UserRingBuffer::new(urb_map).expect("failed to create user ringbuf");
    let err = user_ringbuf
        .reserve::<[u8; 1024 * 1024]>()
        .unwrap_err();
    assert!(err.to_string().contains("Requested size is too large"), "{err:#}");
}
```

jfernandez replied:
Added test cases for reservation too big and not enough space errors

danielocfb commented on libbpf-rs/src/user_ringbuf.rs:
```diff
@@ -0,0 +1,171 @@
+            // Fetch the current value of errno to determine the type of error.
+            let errno = unsafe { *libc::__errno_location() };
+            match errno {
+                E2BIG => Error::with_invalid_data("Requested size is too large"),
+                ENOSPC => Error::with_invalid_data("Not enough space in the ring buffer"),
```
Nit: Can we use lower case error messages, please?
```suggestion
                E2BIG => Error::with_invalid_data("requested size is too large"),
                ENOSPC => Error::with_invalid_data("not enough space in the ring buffer"),
```

danielocfb reviewed (changes requested):
This seems mostly fine. Left a few comments/questions. I am unsure about the whole generic argument `T` thing, though. From what I recall all our existing APIs assume and work with raw buffers. I am not convinced the generic `T` approach is sound, because users could pass any object and we'd just treat it as a byte blob. But that may blow up left and right, depending on what is being passed in. So I'd think we need to switch over to requiring raw buffers here as well. What do you think?

jfernandez commented:
> So I'd think we need to switch over to requiring raw buffers here as well. What do you think?

I'm open to it, but could you give me an example of what how you'd envision the API with raw buffers? 

> because users could pass any object and we'd just treat it as a byte blob. But that may blow up left and right, depending on what is being passed in.

I'm trying to better understand this concern. When users pass in the `T` type in the `reserve` fn, we only use it to figure out how many bytes we need to reserve in the ringbuf, and then to cast the raw pointer. This is what the user would manually if we offered a simpler API. In what ways could this behavior blow up?

arighi commented on libbpf-rs/src/user_ringbuf.rs:
```diff
@@ -0,0 +1,171 @@
+
+        sample.submitted = true;
+
+        // The libbpf function does not return an error code, so we cannot
+        // determine if the submission was successful. We assume that the
+        // submission was successful if the function returns without error.
+        Ok(())
```
I can probably answer this, because I raised the same concern. The idea here is to design the API in a way so that we can potentially add further sanity checks in the future without breaking the API. So, for now, it always returns Ok, but in the future, if we need to return a failure, we can add that and be fully backward compatible.

jfernandez replied:
Yes. Our take is that the libbpf submit API may evolve with error handling.

danielocfb replied:
Okay, I don't feel strongly about making it infallible. That being said, I'd think adjusting the API in `libbpf` would at least constitute a soft breakage and so I am not sure such an change would fly; at the same time `libbpf-rs` doesn't claim to have a stable API yet, so changing it to return a `Result` wouldn't be a biggie (though I appreciate the foresight). In any event, it's a minor thing and either way is fine with me.

jfernandez commented on libbpf-rs/src/user_ringbuf.rs:
```diff
@@ -0,0 +1,171 @@
+    // A non-null pointer to data of type `T` within the ring buffer.
+    ptr: NonNull<T>,
+
+    // `PhantomData<T>` signals to the Rust compiler the ownership and variance
+    // of `T`. This ensures that `UserRingBufferSample` complies with Rust’s
+    // safety rules regarding lifetimes and drop semantics, despite not storing
+    // `T` directly.
+    _marker: PhantomData<T>,
```
This member made more sense in an initial implementation I was working on, but I agree now that it's most likely not needed. I can remove this for now until it's clear it's needed.

jfernandez commented on libbpf-rs/src/user_ringbuf.rs:
```diff
@@ -0,0 +1,171 @@
+    ///
+    /// You can use this method to modify the data within the ring buffer prior
+    /// to submitting the sample.
+    pub fn as_mut(&mut self) -> &mut T {
```
Yeah, possibly. I'm hoping we can agree on whether generics are the right solution for this API before addressing this feedback.

jfernandez replied:
Implemented in the latest push.

danielocfb commented:
> > So I'd think we need to switch over to requiring raw buffers here as well. What do you think?
> 
> I'm open to it, but could you give me an example of how you'd envision the API with raw buffers?

I'd think the most straight forward way would be to adjust `reserve` to accept a usize of the number of bytes to reserve and then have `UserRingBufferSample` deref (or `as_mut`) to `&mut [u8]` of the corresponding size.

I understand that user experience-wise this may not exactly be a step up and we should see how to improve the situation. However, we should make sure not to introduce unsoundness issues or other footguns, while also not diverging from how other APIs are built without a clear path for convergence.

> > because users could pass any object and we'd just treat it as a byte blob. But that may blow up left and right, depending on what is being passed in.
> 
> I'm trying to better understand this concern. When users pass in the `T` type in the `reserve` fn, we only use it to figure out how many bytes we need to reserve in the ringbuf, and then to cast the raw pointer. This is what the user would do manually if we offered a simpler API. In what ways could this behavior blow up?

It may not be a huge deal if objects are only ever used as input. But even that is effectively violated here (see 2) below). A user is free to make `T` anything. Without knowledge of the implementation their is no way for them to gauge behavior and or correctness. Let's say the user has `T` as an object of:
```rust
enum Object {
  Bool(bool),
  Int(usize),
  String(String),
}
```
There is nothing preventing them from doing just that or hinting at the fact that this is problematic. Why is it problematic? I'd guess there may be more, but I see two (and a half) issues:

1) Rust doesn't really define the layout of such objects and even if it did it's a Rust `enum`, which is unlikely to be understood by the kernel. If they map it to a C `union` there will be a mismatch. So that's a footgun. But it's not a soundness issue at least (I believe).

1.1) Closely related, but perhaps noteworthy: for objects that heap allocate (or really anything that has an indirection, I suppose) the API will just copy that indirection. But that's almost (?) never what a user would desire/expect.

2) Let's say we reserve a sample for `Object`. The moment we call `sample.as_mut` we are already in undefined behavior land (i.e., have unsound code): I don't know what's the contract that `user_ring_buffer__reserve` has with respect to the produced pointer, but no matter what the pointed to memory is initialized with, one can come up with cases for which this memory does not constitute, in fact, a valid `T`. But that's illegal from the Rust side of things. The compiler assumes that `&mut T` always references a valid instance of `T`. So *at the very least*, `as_mut` should return a `MaybeUninit<T>` reference to eliminate this unsoundness issue.

These issues are one of the reason why crates such as [`plain`](https://crates.io/crates/plain) or [`pod`](https://crates.io/crates/pod) exist: they move the assurance that a type is valid for any bit pattern (C plain old data) and potentially even the conversion into a byte buffer into the implementation of an `unsafe` trait.

We haven't decided in favor of any of those (or providing our own trait) and so right now it is up to the user how to get the byte buffer. But it's what other APIs in this crate require (grep for `plain` in the code base).

jfernandez commented:
> These issues are one of the reason why crates such as [plain](https://crates.io/crates/plain) or [pod](https://crates.io/crates/pod) exist: they move the assurance that a type is valid for any bit pattern (C plain old data) and potentially even the conversion into a byte buffer into the implementation of an unsafe trait.

> We haven't decided in favor of any of those (or providing our own trait) and so right now it is up to the user how to get the byte buffer. But it's what other APIs in this crate require (grep for plain in the code base).

Thank you for the detailed explanation. I see your concern now and I agree that it may make sense to delegate that responsibility to the user for now.

> I'd think the most straight forward way would be to adjust reserve to accept a usize of the number of bytes to reserve and then have UserRingBufferSample deref (or as_mut) to &mut [u8] of the corresponding size.

I'll give this a try locally. I'll report back or just follow up with an updated PR it I don't find any issues with that proposal.

jfernandez commented:
@danielocfb I removed generics and made `reserve` take the size in bytes. `UserRingBufferSample` will now store the c_void pointer and the size. I implemented `Deref` and `DerefMut` as you suggested, and they return the byte slice. Docs/comments were updated as well. I made the unit test use `plain`.

arighi reviewed (approved):
Just tested the latest code in `scx_rustland` (https://github.com/arighi/scx/tree/dispatch-ringbuf) and it works great!

Transitioning from a BPF queue to a user ring buffer allows to reduce the scheduler's overhead by approximately 30% when the system is over-commissioned (CPU utilization goes from 55% to 40%, in absolute terms).

So, FWIW, tested & approved. Thanks for working on this!

danielocfb reviewed (approved):
Looks great to me, thanks!

danielocfb commented:
Let me know if you guys need a new release cut with this support or if it's fine to consume from git for the time being.

arighi commented:
> Let me know if you guys need a new release cut with this support or if it's fine to consume from git for the time being.

It would be great to have a new release, so that we can point directly to the libbpf-rs crate (pointing to the git repo in `Cargo.toml` makes our packaging more tricky). But IIUC you are already cutting a new release, right?

Thanks for the quick responses and review @danielocfb !

danielocfb commented:
> But IIUC you are already cutting a new release, right?

Yep, Tejun asked about it privately. Should be available by end of day today.

danielocfb commented:
Just released it: https://crates.io/crates/libbpf-rs/0.23.1

jfernandez commented:
Thank you!
